chain/checkpointing/sub.go

```diff
@@ -0,0 +1,506 @@
+			if c.init && c.config != nil {
+
+				data := oldTs.Cids()[0]
+
+				var partyList string = ""
+				for _, partyId := range c.orderParticipantsList() {
+					partyList += partyId + "\n"
+				}
+				hash, err := CreateConfig([]byte(partyList))
+				if err != nil {
+					panic(err)
+				}
+
+				c.CreateCheckpoint(ctx, data.Bytes(), hash)
+			}
+		}
+
+		// Changes detected so generate new key
+		if oldSt.MinerCount != newSt.MinerCount {
+			fmt.Println("Generate new config")
+
+			return true, nil, nil
+		}
+
+		return false, nil, nil
+	}
+
+	err := c.events.StateChanged(checkFunc, changeHandler, revertHandler, 5, 76587687658765876, match)
+	if err != nil {
+		return
+	}
+}
+
+func (c *CheckpointingSub) Start(ctx context.Context) {
+	topic, err := c.pubsub.Join("keygen")
+	if err != nil {
+		panic(err)
+	}
+	c.topic = topic
+
+	// and subscribe to it
+	sub, err := topic.Subscribe()
+	if err != nil {
+		panic(err)
+	}
+	c.sub = sub
+
+	c.listenCheckpointEvents(ctx)
+}
+
+func (c *CheckpointingSub) LoopHandler(ctx context.Context, h protocol.Handler, network *Network) {
+	for {
+		msg, ok := <-h.Listen()
+		if !ok {
+			network.Done()
+			// the channel was closed, indicating that the protocol is done executing.
+			fmt.Println("Should be good")
+			return
+		}
+		network.Send(ctx, msg)
+
+		for _, _ = range network.Parties() {
+			msg = network.Next(ctx)
+			h.Accept(msg)
+		}
+	}
+}
+
+func (c *CheckpointingSub) CreateCheckpoint(ctx context.Context, cp, data []byte) {
+	idsStrings := c.orderParticipantsList()
+	fmt.Println("Participants list :", idsStrings)
+	fmt.Println("Precedent tx", c.ptxid)
+	ids := c.formIDSlice(idsStrings)
+	taprootAddress := PubkeyToTapprootAddress(c.pubkey)
+
+	pubkey := c.config.PublicKey
+	if c.newconfig != nil {
+		pubkey = c.newconfig.PublicKey
+	}
+
+	pubkeyShort := GenCheckpointPublicKeyTaproot(pubkey, cp)
+	newTaprootAddress := PubkeyToTapprootAddress(pubkeyShort)
+
+	if c.ptxid == "" {
+		fmt.Println("Missing precedent txid")
+		taprootScript := GetTaprootScript(c.pubkey)
+		success := AddTaprootScriptToWallet(taprootScript)
+		if !success {
+			panic("failed to add taproot address to wallet")
+		}
+
+		ptxid, err := WalletGetTxidFromAddress(taprootAddress)
+		fmt.Println(taprootAddress)
+		if err != nil {
+			panic(err)
+		}
+		c.ptxid = ptxid
+		fmt.Println("Found precedent txid:", c.ptxid)
+	}
+
+	index := 0
+	value, scriptPubkeyBytes := GetTxOut(c.ptxid, index)
+
+	if scriptPubkeyBytes[0] != 0x51 {
+		fmt.Println("Wrong txout")
+		index = 1
+		value, scriptPubkeyBytes = GetTxOut(c.ptxid, index)
+	}
+	newValue := value - FEE
+
+	payload := "{\"jsonrpc\": \"1.0\", \"id\":\"wow\", \"method\": \"createrawtransaction\", \"params\": [[{\"txid\":\"" + c.ptxid + "\",\"vout\": " + strconv.Itoa(index) + ", \"sequence\": 4294967295}], [{\"" + newTaprootAddress + "\": \"" + fmt.Sprintf("%.2f", newValue) + "\"}, {\"data\": \"" + hex.EncodeToString(data) + "\"}]]}"
```

Comment:
Quick question: why are you defining the raw payload instead of creating a struct and then marshalling? It may be more readable and less prone to bugs.